src/ray/gcs/client_def.h

```diff
@@ -0,0 +1,95 @@
+#ifndef RAY_GCS_CLIENT_DEF_H
+#define RAY_GCS_CLIENT_DEF_H
+
+#include <boost/optional/optional.hpp>
+#include <string>
+#include "ray/gcs/tables.h"
+
+namespace ray {
+
+namespace gcs {
+
+/// \class ClientOption
+/// GCS client's options(configuration items), such as service address, service password.
+class ClientOption {
+ public:
+  /// Constructor of ClientOption.
+  ///
+  /// \param ip GCS service ip
+  /// \param port GCS service port
+  /// \param password GCS service password
+  /// \param is_test_client Is test client
+  ClientOption(const std::string &ip, int port, const std::string &password,
+               bool is_test_client = false)
+      : server_ip_(ip),
+        server_port_(port),
+        password_(password),
+        is_test_client_(is_test_client) {
+#if RAY_USE_NEW_GCS
+    command_type_ = CommandType::kChain;
+#else
+    command_type_ = CommandType::kRegular;
+#endif
+  }
+
+  /// This constructor is only used for testing(RedisGcsClient's test).
+  ///
+  /// \param ip Gcs service ip
+  /// \param port Gcs service port
+  /// \param command_type Command type of RedisGcsClient
+  ClientOption(const std::string &ip, int port, CommandType command_type)
+      : server_ip_(ip),
+        server_port_(port),
+        command_type_(command_type),
+        is_test_client_(true) {}
+
+  // GCS server address
+  std::string server_ip_;
+  int server_port_;
+
+  // Password of GCS server.
+  std::string password_;
+  // GCS command type. If CommandType::kChain, chain-replicated versions of the tables
+  // might be used, if available.
+  CommandType command_type_ = CommandType::kChain;
+
+  // If it's test client.
+  bool is_test_client_{false};
+};
+
+/// \class ClientInfo
+/// Information of GCS client, such as client id.
+/// Only if it's raylet'client, it should contain much more information included by
+/// ClientTableData:
+/// node manager's address, object manager's port and so on.
+class ClientInfo {
+ public:
+  /// Constructor for worker, raylet monitor
+  ClientInfo() : id_(ClientID::FromRandom()) {}
+
+  /// Constructor for worker, raylet monitor
+  explicit ClientInfo(const ClientID &id) : id_(id) {}
+
+  /// Constructor for raylet
+  ClientInfo(const ClientTableData &client_data) : client_data_(client_data) {
+    id_ = ClientID::FromBinary(client_data.client_id());
+  }
+
+  const ClientID &GetClientID() const {
+    RAY_DCHECK(!id_.IsNil());
+    return id_;
+  }
+
+  bool IsRaylet() { return client_data_.is_initialized(); }
```

Comment:
a related question is how are we going to handle the `ClientTable` if it's a worker?

Reply:
Raylet need to send node information (those information contains in ClientTableData) to GCS to register itself. 
If it's worker, you can use "WorkerInfo" instead of ClientTableData. Not implament yet, will do it in next PRs.  Might impl a constructor like: ClientInfo(const WorkerInfo &worker_info). Or make ClientInfo a template class.

If what you means it's how to access data from ClientTable? Not implament too, maybe use NodeStateAccessor to do so.

Reply:
Thanks. Sorry I wasn't clear enough, my question was that `ClientTable` currently makes use of local client ID, but for a worker it's not available, then how would worker uses `ClientTable`?

Reply:
> Thanks. Sorry I wasn't clear enough, my question was that `ClientTable` currently makes use of local client ID, but for a worker it's not available, then how would worker uses `ClientTable`?

I see.  Actually it won't be necessary to pass `client id` to `ClientTable`. `ClientTable` will init `client id` to `nil` inside. If you call `Connect` to register as a `node`(`raylet`), `ClientTable` will update `client id` equal to the input params. If not, `client id` will be `nil`, and you still can call any function you need.